### ISSUE TITLE
SkiaSharp Canvas does not redraw

### ISSUE THREAD

**Issue by GuuD** [fsprojects/Fabulous#182]
**Applies to:** Android, possibly iOS
**Description:**
Regardless of model updates, SkiaSharp Canvas never redraws, and stuck on first drawn image.

I suspect the issue is not calling .InvalidateSurface() method on underlying SKCanvasView, but I don't know enough about Xamarin approaches to be sure that this is idiomatic way of triggering repaints. 

If this is indeed a root cause, should I file pull request that adds .InvalidateSurface() call to update function of SkiaSharp.fs?
```fsharp
let update (prevOpt: ViewElement voption) (source: ViewElement) (target: SKCanvasView) = 
    View.UpdateView (prevOpt, source, target)
    //Add this to trigger redraw on all but first invocations of update
    match prevOpt with
    | ValueSome _ -> target.InvalidateSurface()
    | _ -> ()
    source.UpdatePrimitive(prevOpt, target, CanvasEnableTouchEventsAttribKey, (fun target v -> target.EnableTouchEvents <- v))
    source.UpdatePrimitive(prevOpt, target, IgnorePixelScalingAttribKey, (fun target v -> target.IgnorePixelScaling <- v))
    source.UpdateEvent(prevOpt, PaintSurfaceAttribKey, target.PaintSurface)
    source.UpdateEvent(prevOpt, TouchAttribKey, target.Touch)
```

**Comment by TimLariviere** [comment]
Quickly skimming through SkiaSharp's docs, and seems you're right. MSDN recommends using `InvalidateSurface` to force a redraw -- even though any other `Invalidate*` method should be used really carefully in Xamarin/WPF for performance reasons.

I don't think invalidating the surface at each update is a good idea.
The view gets redrawn for any message sent by your app, and doing so might add flickering on your canvas.

**Example**: You have a page with an Entry and a SKCanvasView. To track what the user inputs, you listen to `textChanged` and send a message each time.
This means that the canvas will get redrawn at each keystroke, even though you don't need to update it.

Maybe we should expose a new parameter that provides with an "Invalidate" action when the control is created?
```fsharp
static member SKCanvasView(?paintSurface: (SKPaintSurfaceEventArgs -> unit),
                           ?touch: (SKTouchEventArgs -> unit),
                           ?enableTouchEvents: bool,
                           ?ignorePixelScaling: bool,
                           ?loaded: ((unit -> unit) -> unit))

/// App

let mutable invalidateCanvas: unit -> unit = ignore

let update msg model =
     match msg with
     | Touched ->
         invalidateCanvas()
         model

let view model dispatch =
     View.SKCanvasView(
         touch = (fun e -> dispatch Touched),
         loaded = (fun f -> invalidateCanvas <- f)
      )
```

@dsyme What do you think?

**Comment by dsyme** [comment]
This would be ok. But what about 
```
static member SKCanvasView(?paintSurface: (SKPaintSurfaceEventArgs -> unit),
                           ?touch: (SKTouchEventArgs -> unit),
                           ?enableTouchEvents: bool,
                           ?ignorePixelScaling: bool,
                           ?invalidate: bool)

/// App

let update msg model =
     match msg with
     | Touched ->
         { model with InvalidateCanvas = true }

let view model dispatch =
     View.SKCanvasView(
         touch = (fun e -> dispatch Touched),
         invalidate = model.InvalidateCanvas
      )
```
You'd basically have to recompute InvalidateCanvas at every update (to make sure it gets set back to false).

Are there other invalidation techniques, e.g. region based?  Those would be handled similarly I suppose?

**Comment by TimLariviere** [comment]
Your solution is better, more aligned with Elmish.
>You'd basically have to recompute InvalidateCanvas at every update (to make sure it gets set back to false).

I think dispatching a message to reset InvalidateCanvas when `paintSurface` is triggered should be enough, as this one is called immediately after `InvalidateSurface()`.
We could also expose a new `invalidated` callback to be more explicit.

**Comment by dsyme** [comment]
Sounds like a plan

**Comment by dsyme** [comment]
@TimLariviere I assume the `invalidate` property would only trigger InvalidateSurface once when its value changes false --> true?  e.g. implemented by

                source.UpdatePrimitive(prevOpt, target, InvalidateAttribKey, (fun target v -> if v then target.InvalidateSurface()))

**Comment by TimLariviere** [comment]
Yes, that what makes the most sense.

Although `true --> true` could also be valid.
With only `false --> true`, if someone wants to invalidate every time, they would have to constantly dispatch 2 messages to force invalidation.

**Comment by dsyme** [comment]
I believe this to be fixed by #190 though have not created a specific test for it

We should double-check manually and add some docs to the SkiaSharp extension docs
